Fixes #2990 - Optimize some calls to the TaskTracker

kolloch opened the pull request:
* by making less of them
* by using asynchronous calls to avoid blocking threads

I also made the AppDataMap and App part of the TaskTracker interface to
avoid conversion and provide richer querying syntax.

kolloch commented:
rebased

gkleiman commented on src/main/scala/mesosphere/marathon/core/appinfo/impl/AppInfoBaseData.scala:
```diff
@@ -48,6 +50,11 @@ class AppInfoBaseData(
     }
   }
 
+  lazy val tasksMap: Future[TaskTracker.AppDataMap] = {
```
wouldn't tasksMapFuture be more consistent with the rest of the names in this class?

kolloch replied:
Yes, thanks :+1:

gkleiman commented on src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala:
```diff
@@ -35,5 +35,69 @@ trait TaskTracker {
+
+    def empty: AppDataMap = AppDataMap(Map.empty)
+  }
+  /**
```
We normally add an empty line between methods

kolloch replied:
Ok

gkleiman commented on src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala:
```diff
@@ -35,5 +35,69 @@ trait TaskTracker {
+  /**
+    * Contains all tasks grouped by app ID.
+    */
+  case class AppDataMap private (appTasks: Map[PathId, TaskTracker.App]) {
```
I guess it is outside of the scope of this PR, but it would be nice at some point to rename `AppDataMap` to `TasksByApp`.

kolloch replied:
I added that to https://github.com/mesosphere/marathon/issues/3003, so we have a pure renaming PR.

kolloch commented:
* Renamed tasksMap to tasksMapFuture.
* Added newline between methods.

```diff
diff --git a/src/main/scala/mesosphere/marathon/core/appinfo/impl/AppInfoBaseData.scala b/src/main/scala/mesosphere/marathon/core/appinfo/impl/AppInfoBaseData.scala
index 4423808..2dc0b37 100644
--- a/src/main/scala/mesosphere/marathon/core/appinfo/impl/AppInfoBaseData.scala
+++ b/src/main/scala/mesosphere/marathon/core/appinfo/impl/AppInfoBaseData.scala
@@ -50,7 +50,7 @@ class AppInfoBaseData(
     }
   }

-  lazy val tasksMap: Future[TaskTracker.AppDataMap] = {
+  lazy val tasksMapFuture: Future[TaskTracker.AppDataMap] = {
     log.debug("Retrieve tasks")
     taskTracker.listAsync()
   }
@@ -86,7 +86,7 @@ class AppInfoBaseData(
   private[this] class AppData(app: AppDefinition) {
     lazy val now: Timestamp = clock.now()

-    lazy val tasksFuture: Future[Iterable[MarathonTask]] = tasksMap.map(_.getTasks(app.id))
+    lazy val tasksFuture: Future[Iterable[MarathonTask]] = tasksMapFuture.map(_.getTasks(app.id))

     lazy val healthCountsFuture: Future[Map[String, Seq[Health]]] = {
       log.debug(s"retrieving health counts for app [${app.id}]")
@@ -132,7 +132,7 @@ class AppInfoBaseData(

       log.debug(s"assembling rich tasks for app [${app.id}]")

-      val tasksByIdFuture = tasksMap.map(_.appTasks.get(app.id).map(_.taskMap).getOrElse(Map.empty))
+      val tasksByIdFuture = tasksMapFuture.map(_.appTasks.get(app.id).map(_.taskMap).getOrElse(Map.empty))
       val healthStatusesFutures = healthCheckManager.statuses(app.id)
       for {
         tasksById <- tasksByIdFuture
diff --git a/src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala b/src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala
index b419b82..e7bf6b8 100644
--- a/src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala
@@ -74,6 +74,7 @@ object TaskTracker {
     def of(appTasks: collection.immutable.Map[PathId, TaskTracker.App]): AppDataMap = {
       new AppDataMap(appTasks.withDefault(appId => TaskTracker.App(appId)))
     }
+
     def of(apps: TaskTracker.App*): AppDataMap = of(Map(apps.map(app => app.appId -> app): _*))

     def empty: AppDataMap = AppDataMap(Map.empty)
```

gkleiman commented:
LGTM, will merge as soon as the tests pass.

gkleiman commented:
Merged :rocket:
